File: SubwayMon/csv-parser.cpp

#include "csv-parser.h"

#include <sstream>

std::vector<std::vector<std::string>> parseCSV(const std::string& input) {
  std::vector<std::vector<std::string>> result;

  std::istringstream stream { input };

  while (stream.good()) {
    constexpr int maxLineSize = 2048;
    char buffer[maxLineSize];
    stream.getline(buffer, maxLineSize);

    std::vector<std::string> fields {};
    std::ostringstream field {};
    auto* ptr = buffer;
    bool inQuoted = false;

    while (*ptr) {
      if (inQuoted) {
        if (*ptr == '"') {
          inQuoted = false;
        } else if (*ptr == '\\' && *(ptr + 1) == '"') {
          field << '"';
          ++ptr;
        } else {
          field << *ptr;
        }
      } else {
        if (*ptr == ',') {
          fields.push_back(field.str());
          field.str("");
        } else if (*ptr == '"') {
          inQuoted = true;
        } else {
          field << *ptr;
        }
      }

      ++ptr;
    }

    result.push_back(fields);
  }

  return result;
}

File: SubwayMon/csv-parser.swift
//
//  csv-parser.swift
//  SubwayMon
//
//  Created by Owen Yamauchi on 3/4/18.
//  Copyright © 2018 Owen Yamauchi. All rights reserved.
//

func parseCsv(_ input: String) -> Array<Array<String>> {
  var result = Array<Array<String>>()

  input.enumerateLines { line, _ in
    var fields = Array<String>()
    var field = String()
    var inQuoted = false
    var index = line.startIndex

    while index < line.endIndex {
      if inQuoted {
        if line[index] == "\"" {
          inQuoted = false
        } else if line[index] == "\\" && line[line.index(after: index)] == "\"" {
          field.append("\"")
          index = line.index(after: index)
        } else {
          field.append(line[index])
        }
      } else {
        switch line[index] {
        case ",":
          fields.append(field)
          field = String()
        case "\"":
          inQuoted = true
        default:
          field.append(line[index])
        }
      }

      index = line.index(after: index)
    }

    result.append(fields)
  }

  return result
}
